### ISSUE TITLE
Optimization: Exclude non-sort/filter relevant fields in reduce-extraction-to-projection

### ISSUE THREAD

**Issue by Yogu** [arangodb/arangodb#7821]
## My Environment

* __ArangoDB Version__:        3.4
* __Storage Engine__:             RocksDB
* __Deployment Mode__:       Single Server
* __Deployment Strategy__:   Manual Start
* __Infrastructure__:               local
* __Operating System__:        Windows 10

## Component, Query & Data

__Affected feature__: AQL query optimizer

__AQL query (if applicable)__:

```aql
for m in businessMessages
filter m.tray == "IN"
sort m.updatedAt 
limit 10
let d = m//DOCUMENT(m._id)
return [d.metadata, d.createdAt, d.updatedAt, d._key, d._id]
```

__AQL explain (if applicable)__:

```
Query String:
 for m in businessMessages
 filter m.tray == "IN"
 sort m.updatedAt 
 limit 10
 let d = m//DOCUMENT(m._id)
 return [d.metadata, d.createdAt, d.updatedAt, d._key, d._id]

Execution plan:
 Id   NodeType                   Est.   Comment
  1   SingletonNode                 1   * ROOT
  2   EnumerateCollectionNode   25467     - FOR m IN businessMessages   /* full collection scan */
  3   CalculationNode           25467       - LET #2 = (m.`tray` == "IN")   /* simple expression */   /* collections used: m : businessMessages */
  4   FilterNode                25467       - FILTER #2
  5   CalculationNode           25467       - LET #4 = m.`updatedAt`   /* attribute expression */   /* collections used: m : businessMessages */
  6   SortNode                  25467       - SORT #4 ASC
  7   LimitNode                    10       - LIMIT 0, 10
  9   CalculationNode              10       - LET #6 = [ m.`metadata`, m.`createdAt`, m.`updatedAt`, m.`_key`, m.`_id` ]   /* simple expression */   /* collections used: m : businessMessages */
 10   ReturnNode                   10       - RETURN #6

Indexes used:
 none

Optimization rules applied:
 Id   RuleName
  1   move-calculations-up
  2   move-filters-up
  3   remove-unnecessary-calculations
  4   move-calculations-up-2
  5   move-filters-up-2
```

__Dataset__:

25 000 documents in "businessMessages" which are rather large (contain a long text field)

## Steps to reproduce

__Problem__:

The given query takes a lot of memory to execute because it loads 25 000 full documents in memory, sorts them, and then only returns 10.

__Expected result__:

If I add an indirection as indicated in the comment in the AQL statement (`let d = DOCUMENT(m._id)`), the query executes ~4 times faster and takes a lot less memory - just because the `reduce-extraction-to-projection` optimization kicks in and reduces the document size in memory significantly. In the original query, this optimization is not applied because more than 5 fields are accessed. However, those are only accessed in the RETURN clause, so it's way more efficient to just fetch them for the 10 documents we requested instead of for all of them.

Thus, the expection would be that ArangoDB sees the filter will return a lot more documents than the RETURN statement will process and in turn introduces this indirection I added in the query automatically.

**Comment by dothebart** [comment]
To me this sounds much of a duplicate of @jsteemann explained here: https://github.com/arangodb/arangodb/issues/7782#issuecomment-448384616

**Comment by Yogu** [comment]
@dothebart that issue / comment actually describes a different problem I currently encounter, so thanks for the link. However, I don't see how it is related to this issue - I'm not using or intending to use an index here.

**Comment by Yogu** [comment]
I don't have a problem with ArangoDB looking at the document. My problem is that it loads the full documents into memory and not only the fields that are needed for sorting. More specifically, that the `reduce-extraction-to-projection` optimization does not apply.

To explain our use case: We have a million largish documents (a few kilobytes) which we sometimes need to sort without an index, and then return the first 10 items. When the query in total only reads 5 fields of the documents, it takes a few seconds, but finishes fine. When the query reads more than 5 fields of the documents, we run out of memory because basically the whole collection is read into memory and sorted there. This can be "fixed" by introducing the indirection via `DOCUMENT(m._id)` I mentioned in the issue.

My suggestion would be that ArangoDB should do this indirection automatically if the LIMIT value is significantly lower than the collection length.

A whole different question is why ArangoDB even needs to keep all the documents in memory for the sort if we're only intersted in the top 10 documents anyway. I asked that on slack yesterday (https://arangodb-community.slack.com/archives/C5L3P6SUF/p1545319824123700), I can create a separate issue for that if it's not just something I'm doing wrong. Anyway, I think this issue here would still apply because I would guess that a projected index scan is still faster than one that loads the whole documents, regardless of the memory usage.

But maybe I'm jumping wrong conclusions here, so please correct me if I'm wrong. I'm trying to understand what ArangoDB can do and where the limitations are. By the way, we (AEB) now also bought developer support from January onwards, so maybe that would provide a better framework for that.

**Comment by dothebart** [comment]
Ok, let me explain a bit about the basics here first. 

With document stores vs. table based stores you have the tradeoff that the document is stored as a blob, so the position of random attributes actually is random; while the table-store knows the definitive place on disk where one row of a column can be located. 
This means that in order to access an attribute of a document you have to parse the document. Indices keep a copy of the attribute they index, so they don't have to re-parse. 

Now another part of the Basics - AQL implementation. AQL is implemented as a stream of operators, similary as you would do on a unix shell with pipes if you use sed and grep. So, Documents come in at the left, are reduced by filters, or sub-attribute accesses and no further use of the document. 

The optimizer now tries to re-order this pipe stream to make it as efficient as possible, to reduce the number of documents as quick as possible in the chain. One way to achieve this is instead of a full collection scan and manually filter have an index used to only give you as many documents as you need. 

The current behaviour of the sort block is to keep all documents, sort them, return as many as demanded. 
This is pretty simple, and quick - can use ready made sort algorithms etc. 

Now one could add an optimizer rule that inspects LIMIT nodes, copies it over to the sort node, so it may only keep `skip+limit` documents, pull in documents, push documents out that are over the limit, throw pulled documents away instantly if they are above the limit. While this may be more memory efficient, its most probably not faster in most cases.

Plus in your case you do a full collection scan, so it _will_ still load all documents, parse them, etc., so this can't become fast either. 

So, there may be other solutions, like splitting the documents, putting the big parts into another collection, etc, or even the `document()` approach you took.

**Comment by mpoeter** [comment]
@dothebart I agree with almost everything you wrote, but I do think that this would be a very useful optimization that would not only reduce memory requirements but also improve performance - at least in cases where the limit is significantly smaller than the collection size (which I would consider a very frequent case).

Let's assume we have a collection with *n* documents and a LIMIT *k*. Of course you still have to perform a full scan and load+parse all *n* documents, but you only have to perform a partial sort and keep *k* documents. This could be efficiently done using a heap with a fixed size *k*. An update of the heap has complexity *O(* log *k)*, so the total complexity would be *O(n* log *k)*, while a sort of all documents would have complexity *O(n* log *n)*.

**Comment by jsteemann** [comment]
@Yogu : All the above comments aside, I think the situation can also be improved by adding a combined index on the two fields `["tray", "updatedAt"]`. This should also make the query a lot faster without any change in the server-side code.

In addition I think that @mpoeter 's suggestion sounds very sensible, because it should also help in most situations.

**Comment by Yogu** [comment]
@jsteemann yes obviously an index would avoid this problem, but it would be good if the database would not blow up on queries that don't use an index :)

**Comment by jsteemann** [comment]
@yogu: the problem here is that this is also a kind-of-difficult decision for the optimizer.

The "reduce-extraction-to-projection" rule is currently limited to 5 attributes code-wise, and the above query uses more attributes from the document, so it does not kick in.
While the threshold for the rule could easily be adjusted to any other number of attributes, it is not 100% clear if increasing the number of attributes may make some queries perform worse.

For example, let's say we increased the threshold to 8 attributes. Now, this particular query may be fine, but queries that deal with small documents (small document size, few attributes contained in the documents) may actually get slower after this change, because the optimization may lead to the individual document attributes being copied out of the documents one by one - which in sum may be slower than copying out the entire document in one go.

Still assuming a threshold of 8, when encountering a document with just 8 attributes would make the rule search for the desired 8 attributes inside the document and then copy the 8 attributes out one by one. That means 8 binary searches plus 8 times memcpy. Whereas without the optimization it would mean no binary searches and just a simple memcpy.

That means the optimization may decrease performance in the worst case. However, it should never do any harm in terms of memory usage (in the worst case, the memory usage is the same as without the optimization kicking in). And in the best case, the optimization may save a lot of memory. 

Obviously the sweet spot depends a lot on how big the documents are and how many attributes are contained in them. The problem is that the optimizer doesn't know anything about this when planning queries, so its assumptions must be somewhat generic and conservative.
That leaves the question if there are some additional cases in which we can more or less safely apply the "reduce-extraction-to-projection" rule without making the performance of simple/small document queries worse.

Apart from that, I think @mpoeter's suggestion is still good here, because it will help in many cases to keep memory usage down. I can't yet estimate its performance effects for real-world queries, but it can be given a try.

**Comment by Yogu** [comment]
@jsteemann I agree with everything you wrote. Should I close this issue and open a new one for the limited-size-heap sort as explained by @mpoeter? That seems to be the way to go which would invalidate the need the *reduce-extraction-to-projection* optimization.

To be clear, @mpoeter's suggestion is what I actually had in mind earlier, but I first asked it in slack instead of opening an issue because I was not sure if this optimization really was not already in place. I think it is really important as it reduces the size complexity of memory usage from linear to constant, which is a big deal. Also, it unblocks #7824 which would reduce the time complexity from linear to constant in other cases.

**Comment by jsteemann** [comment]
@Yogu : ok, I am happy if you close this issue and open a more specific one for the heap sort!
Thanks!

**Comment by Yogu** [comment]
Closed, opened #7896 for the heap size limitation.
